lib/lambda_ethereum_consensus/store/state_store.ex

```diff
@@ -7,43 +7,54 @@ defmodule LambdaEthereumConsensus.Store.StateStore do
   alias Types.BeaconState
 
   @state_prefix "beacon_state"
+  @state_block_prefix "beacon_state_by_state"
   @stateslot_prefix @state_prefix <> "slot"
 
-  @spec store_state(BeaconState.t(), Types.root()) :: :ok
+  @spec store_state(BeaconState.t()) :: :ok
   def store_state(%BeaconState{} = state) do
     # NOTE: due to how SSZ-hashing works, hash(block) == hash(header)
-    store_state(state, Ssz.hash_tree_root!(state.latest_block_header))
-  end
-
-  def store_state(%BeaconState{} = state, block_root) do
+    block_root = Ssz.hash_tree_root!(state.latest_block_header)
+    state_root = Ssz.hash_tree_root!(state)
     {:ok, encoded_state} = Ssz.to_ssz(state)
 
-    key = state_key(block_root)
-    Db.put(key, encoded_state)
+    key_block = state_key(block_root)
+    key_state = block_key(state_root)
+    Db.put(key_block, encoded_state)
+    Db.put(key_state, key_block)
 
     # WARN: this overrides any previous mapping for the same slot
-    slothash_key = root_by_slot_key(state.slot)
-    Db.put(slothash_key, block_root)
+    slothash_key_block = root_by_slot_key(state.slot)
+    Db.put(slothash_key_block, block_root)
+  end
+
+  defp get_block_root(root, type) do
+    if type == :state_root do
+      with {:ok, bin} <- root |> block_key() |> Db.get() do
+        bin
+      end
+    else
+      root
+    end
   end
 
-  @spec get_state(Types.root()) ::
+  @spec get_state(Types.root(), :block_root | :state_root) ::
           {:ok, BeaconState.t()} | {:error, String.t()} | :not_found
-  def get_state(root) do
-    with {:ok, bin} <- root |> state_key() |> Db.get() do
+  def get_state(root, root_type) do
+    with {:ok, bin} <- root |> get_block_root(root_type) |> state_key() |> Db.get() do
```

Comment:
For API simplicity, we should have two functions for this: one for `:block_root` and another for `:state_root` and leave this function private.